piptools/cache.py

```diff
@@ -49,7 +57,10 @@ class DependencyCache(object):
     Where X.Y indicates the Python version.
     """
 
+    _cache = None  # type: Dict[str, dict]
```

Comment:
Why did you move it from instance to class?

Reply:
It fixes the following complainings:

```
piptools/cache.py:82: error: Incompatible return value type (got "None", expected "Dict[str, Dict[Any, Any]]")
piptools/cache.py:109: error: Incompatible types in assignment (expression has type "Dict[str, Dict[Any, Any]]", variable has type "None")
piptools/cache.py:111: error: Incompatible types in assignment (expression has type "Dict[<nothing>, <nothing>]", variable has type "None")
piptools/cache.py:122: error: Incompatible types in assignment (expression has type "Dict[<nothing>, <nothing>]", variable has type "None")
```

Reply:
But you completely changed the semantics with this move, which may cause unwanted side-effects. Why didn't you just add an annotation in the initializer?

Reply:
I've tried at first:

```python
self._cache = None  # type: Dict[str, dict]
````
but
```
piptools/cache.py:71: error: Incompatible types in assignment (expression has type "None", variable has type "Dict[str, Dict[Any, Any]]")
```
then 
```python
self._cache = None  # type: Optional[Dict[str, dict]]
```
but
```
piptools/cache.py:82: error: Incompatible return value type (got "Optional[Dict[str, Dict[Any, Any]]]", expected "Dict[str, Dict[Any, Any]]")
```

I've found the solution and now can't find a reference since I made changes a month ago. I should have fixed it in a separate well-described commit with the reference 😒

Reply:
> which may cause unwanted side-effects.

Yeah, now `_cache` is in `DependencyCache.__dict__`.

Reply:
Do you have any thoughts how to fix this and keep self._cache initialized in the constructor?

Reply:
There might a design problem also.

Reply:
```python
from typings import Optional, Dict

class Cache:
    def __init__(self):
        # type: () -> None
        self._cache = None  # type: Optional[Dict[str, dict]]

    @property
    def cache(self):
        # type: () -> Dict[str, dict]
        if self._cache is None:
            self.read_cache()
        return self._cache

    def read_cache(self):
        # type: () -> None
        self._cache = {"cache": {}}
```

```bash
$ mypy t.py
Success: no issues found in 1 source file
```

🤷‍♂

Reply:
I've refactored `read_cache` to assign explicitly `self._cache` and it works
```diff
diff --git piptools/cache.py piptools/cache.py
index 5c5fb34..24327f1 100644
--- piptools/cache.py
+++ piptools/cache.py
@@ -57,7 +57,6 @@ class DependencyCache(object):
     Where X.Y indicates the Python version.
     """

-    _cache = None  # type: Dict[str, dict]

     def __init__(self, cache_dir=None):
         # type: (Optional[str]) -> None
@@ -69,6 +68,7 @@ class DependencyCache(object):
         cache_filename = "depcache-py{}.json".format(py_version)

         self._cache_file = os.path.join(cache_dir, cache_filename)
+        self._cache = None  # type: Optional[Dict[str, dict]]

     @property
     def cache(self):
@@ -78,7 +78,7 @@ class DependencyCache(object):
         lazily loads the cache from disk.
         """
         if self._cache is None:
-            self.read_cache()
+            self._cache = self.read_cache()
         return self._cache

     def as_cache_key(self, ireq):
@@ -103,12 +103,12 @@ class DependencyCache(object):
         return name, "{}{}".format(version, extras_string)

     def read_cache(self):
-        # type: () -> None
+        # type: () -> Dict[str, dict]
         """Reads the cached contents into memory."""
         if os.path.exists(self._cache_file):
-            self._cache = read_cache_file(self._cache_file)
+            return read_cache_file(self._cache_file)
         else:
-            self._cache = {}
+            return {}

     def write_cache(self):
         # type: () -> None
```


```bash
$ mypy .
Success: no issues found in 44 source files
```

Reply:
Is this approach okay?

Reply:
Yeah, looks good.